File: OOP/Encapsulation/Java/Bank_Account.java
package OOP.Encapsulation.Java;

public class Bank_Account {
    private double balance;

    public Bank_Account(double initial_balance) {
        if (initial_balance < 0) {
            throw new IllegalArgumentException("Initial balance cannot be negative");
        }
        this.balance = initial_balance;
    }

    public void deposit(double amount) {
        if (amount <= 0) {
            throw new IllegalArgumentException("Deposit amount must be positive");
        }
        balance += amount;
    }

    public void withdraw(double amount) {
        if (amount <= 0) {
            throw new IllegalArgumentException("Withdraw amount must be positive");
        }
        if (amount > balance) {
            throw new IllegalStateException("Insufficient balance");
        }
        balance -= amount;
    }

    public double get_balance() {
        return balance;
    }
}

File: OOP/Encapsulation/C++/BankAccount.cpp
#include "BankAccount.h"
#include <stdexcept>

BankAccount::BankAccount(double initialBalance) {
  if (initialBalance < 0)
    throw std::invalid_argument("Initial balance cannot be negative");
  balance = initialBalance;
}

void BankAccount::deposit(double amount) {
  if (amount <= 0)
    throw std::invalid_argument("Deposit must be positive");
  balance += amount;
}

void BankAccount::withdraw(double amount) {
  if (amount <= 0)
    throw std::invalid_argument("Withdraw must be positive");
  if (amount > balance)
    throw std::runtime_error("Insufficient balance");
  balance -= amount;
}

double BankAccount::getBalance() const { return balance; }
